src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java

```diff
@@ -0,0 +1,352 @@
+package com.google.research.bleth.simulator;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * An abstract class representing a BLETH simulation.
+ * Can be either a Tracing simulation or a Stalking simulation.
+ */
+public abstract class AbstractSimulation {
+
+    private final String id;
+    private int currentRound;
+    private final int maxNumberOfRounds;
+    private RealBoard board;
+    protected final ImmutableList<Beacon> beacons;
+    protected final ImmutableList<Observer> observers;
+    private IResolver resolver;
+    private final double radius;
+    private HashMap<String, Double> stats = new HashMap<>();
+
+    /** Returns the simulation's real board. */
+    RealBoard getBoard() {
+        return board;
+    }
+
+    /** Run entire simulation logic, including writing data to db. */
+    public void run() {
+        if (currentRound == 0) {
+            writeRoundState(); // round 0 is the initial simulation state
+            currentRound++;
+        }
+        while (currentRound <= maxNumberOfRounds) {
+            moveAgents();
+            updateObserversAwaknessState();
+            beaconsToObservers();
+            observersToResolver();
+            resolverEstimate();
+            writeRoundState();
+            updateSimulationStats();
+            currentRound++;
+        }
+        writeSimulationStats();
+    }
+
+    /** Move all agents according to their movement strategies and update the real board. */
+    void moveAgents() {
+        for (Beacon beacon : beacons) {
+            beacon.move();
+        }
+        for (Observer observer : observers) {
+            observer.move();
+        }
+    }
+
+    /** Update all observers awakeness states according to their awakeness strategies. */
+    void updateObserversAwaknessState() {
+        for (Observer observer : observers) {
+            observer.updateAwakenessState(currentRound);
+        }
+    }
+
+    /**
+     * Pass transmissions from beacons to observers while taking into consideration world-physics parameters,
+     * such as probability of transmission and distance between beacons and observers.
+     */
+    void beaconsToObservers() {
+        for (Beacon beacon : beacons) {
+            for (Observer observer : observers) {
+                if (observer.isAwake()) {
+                    double distance = distance(beacon.getLocation(), observer.getLocation());
+                    if (distance <= radius) {
+                        observer.observe(beacon.transmit());
```

Comment:
Consider separating the transmission from the observation, i.e. ask the beacon to transmit once and then send the transmission to the relevant observers. It represents reality more accurately and the method transmit may contain more logic in the future.